kratos/utilities/model_part_graph_utilities.cpp

```diff
@@ -0,0 +1,262 @@
+    ModelPart::NodesContainerType& rNodes,
+    const Variable<double>& rVar,
+    const DenseVector<double>& colors,
+    const ModelPartGraphUtilities::IndexType ncolors
+)
+{
+    KRATOS_ERROR_IF(rNodes.size() != colors.size()) << "mismatch in the size of rNodes and of colors" << std::endl;
+    //count the fixed entries for every color
+    DenseVector<ModelPartGraphUtilities::IndexType> v(ncolors);
+    v.clear();
+    std::vector<ModelPartGraphUtilities::IndexType> fixed_ids;
+
+    //count the fixed nodes
+    for(ModelPartGraphUtilities::IndexType i=0; i<rNodes.size(); ++i)
+    {
+        if(colors[i]>=0 && (rNodes.begin()+i)->IsFixed(rVar))
+        {
+            v[colors[i]] += 1;
+        }
+    }
+
+    //now we need to fix one node per each color (unless there are some fixed)
+    for(ModelPartGraphUtilities::IndexType i=0; i<rNodes.size(); ++i)
+    {
+        int color = colors[i];
+        if(color>=0 && v[color]==0)
+        {
+            (rNodes.begin()+i)->Fix(rVar);
+            fixed_ids.push_back((rNodes.begin()+i)->Id());
+            v[color] += 1;
+        }
+    }
+
+    return fixed_ids;
+}
+
+void ModelPartGraphUtilities::BreadthFirstSearch(
+    const int startVertex,
+    const int color,
+    const DenseVector<ModelPartGraphUtilities::IndexType>& rRowIndices,
+    const DenseVector<ModelPartGraphUtilities::IndexType>& rColIndices,
+    std::unordered_map<ModelPartGraphUtilities::IndexType, int>& visited)
+{
+    std::queue<int> q;
+    visited.find(startVertex)->second = color;
+    q.push(startVertex);
+
+    while (!q.empty())
+    {
+        int currVertex = q.front();
+        q.pop();
+        for (ModelPartGraphUtilities::IndexType i = rRowIndices[currVertex]; i != rRowIndices[currVertex+1]; ++i)
+        {
+            int adjVertex = rColIndices[i];
+            auto& item = visited.find(adjVertex)->second;
+            if (item==-1)
+            {
+                item=color;
+                q.push(adjVertex);
+            }
+        }
+    }
+}
+
+void ModelPartGraphUtilities::BreadthFirstSearch_ActiveNodesCheck(
+    const int startVertex,
+    const int color,
+    const DenseVector<ModelPartGraphUtilities::IndexType>& rRowIndices,
+    const DenseVector<ModelPartGraphUtilities::IndexType>& rColIndices,
+    std::unordered_map<ModelPartGraphUtilities::IndexType, int>& visited,
+    const std::unordered_map<ModelPartGraphUtilities::IndexType, bool>& active_nodes)
+{
+    std::queue<int> q;
+
+    visited.find(startVertex)->second = color;
+    q.push(startVertex);
+
+    while (!q.empty())
+    {
+        int currVertex = q.front();
+        q.pop();
+        for (ModelPartGraphUtilities::IndexType i = rRowIndices[currVertex]; i != rRowIndices[currVertex+1]; ++i)
+        {
+            int adjVertex = rColIndices[i];
+            auto& item = visited.find(adjVertex)->second;
+            if (item==-1 && active_nodes.find(adjVertex)->second)
```

Comment:
What is statistically more common, the first or the second condition?, the most probable to be false should be the first